python/demo/gsoc2022/demo_sbc.py

```diff
@@ -0,0 +1,244 @@
+r.interpolate(radial_distance)
+
+# definition of Trial and Test functions
+Es = ufl.TrialFunction(V)
+Vs = ufl.TestFunction(V)
+
+# definition of 3d fields for cross and curl operations
+Es_3d = from_2d_to_3d(Es)
+Vs_3d = from_2d_to_3d(Vs)
+
+# Measures for subdomains
+dAu = ufl.Measure("dx", mesh, subdomain_data=cell_tags, subdomain_id=au_tag)
+dBkg = ufl.Measure("dx", mesh, subdomain_data=cell_tags, subdomain_id=bkg_tag)
+dsbc = ufl.Measure("ds", mesh, subdomain_data=facet_tags, subdomain_id=boundary_tag)
+dDom = dAu + dBkg
+
+# normal to the boundary
+n = FacetNormal(mesh)
+n_3d = from_2d_to_3d(n)
+
+# definition of relative permittivity for Au @400nm
+reps_au = -1.0782
+ieps_au = 5.8089
+eps_au = reps_au + ieps_au * 1j
+
+# definition of the relative permittivity over the whole domain
+D = fem.FunctionSpace(mesh, ("DG", 0))
+eps = fem.Function(D)
+au_cells = cell_tags.indices[cell_tags.values == au_tag]
+bkg_cells = cell_tags.indices[cell_tags.values == bkg_tag]
+eps.x.array[au_cells] = np.full(len(au_cells), reps_au + ieps_au * 1j)
+eps.x.array[bkg_cells] = np.full(len(bkg_cells), eps_bkg)
+
+# weak form
+F = - inner(curl(Es), curl(Vs)) * dDom \
+    + eps * k0**2 * inner(Es, Vs) * dDom \
+    + k0**2 * (eps - eps_bkg) * inner(Eb, Vs) * dDom \
+    + (1j * k0 * n_bkg + 1 / (2 * r)) * inner(cross(Es_3d, n_3d),
+                                              cross(Vs_3d, n_3d)) * dsbc  # scattering boundary condition
+
+# splitting in left-hand side and right-hand side
+a, L = lhs(F), rhs(F)
+
+problem = fem.petsc.LinearProblem(a, L, bcs=[], petsc_options={"ksp_type": "preonly", "pc_type": "lu"})
+Eh = problem.solve()
+
+# total electric field E = Es + Eb
+E = fem.Function(V, dtype=np.complex128)
+E.x.array[:] = Eb.x.array[:] + Eh.x.array[:]
+
+# ||E||
+norm_func = sqrt(inner(Eh, Eh))
+V_normEh = fem.FunctionSpace(mesh, lagr_el)
+norm_expr = fem.Expression(norm_func, V_normEh.element.interpolation_points)
+normEh = fem.Function(V_normEh)
+normEh.interpolate(norm_expr)
+
+# save the fields as xdmf files
+save_as_xdmf("data/Es.xdmf", mesh, Eh)
+save_as_xdmf("data/E.xdmf", mesh, E)
+save_as_xdmf("data/normEh.xdmf", mesh, normEh)
+
+# calculation of analytical efficiencies
+q_abs_analyt, q_sca_analyt, q_ext_analyt = calculateAnalyticalEfficiencies(reps_au, ieps_au, n_bkg, wl0, radius_wire)
+
+# vacuum impedance
+Z0 = np.sqrt(mu_0 / epsilon_0)
+
+# magnetic field H
+Hh_3d = -1j * curl_2d(Eh) / Z0 / k0 / n_bkg
+
+Eh_3d = from_2d_to_3d(Eh)
+E_3d = from_2d_to_3d(E)
+
+# intensity of the electromagnetic fields I0 = 0.5*E0**2/Z0
+# E0 = np.sqrt(ax**2 + ay**2) = 1, see background_electric_field
+I0 = 0.5 / Z0
+
+# geometrical cross section of the wire
+gcs = 2 * radius_wire
+
+# quantities for the calculation of efficiencies
+P = 0.5 * inner(cross(Eh_3d, conj(Hh_3d)), n_3d)
+Q = 0.5 * ieps_au * k0 * (inner(E_3d, E_3d)) / Z0 / n_bkg
+
+# normalized efficiencies
+q_abs_fenics_proc = ufl.real(fem.assemble_scalar(fem.form(Q * dAu)) / gcs / I0)
```

Comment:
I would suggest using `np.float64` here, and not `ufl.real`

Reply:
I used `().real`.